File: cplusplus/218TheSkylineProblem.cpp
class Solution {
public:
    vector<pair<int, int>> getSkyline(vector<vector<int>>& buildings) {
        vector<pair<int, int>> heights;
        for (int i = 0; i < buildings.size(); ++i) {
            heights.push_back(make_pair(buildings[i][0], buildings[i][2]));
            heights.push_back(make_pair(buildings[i][1], -buildings[i][2]));
        }
        
        sort(heights.begin(), heights.end());
        
        vector<pair<int, int>> res;
        map<int, int> m;
            
        int max = 0;
        for (int i = 0; i < heights.size(); ++i) {
            if (heights[i].second > 0) {
                if (heights[i].second > max) {
                    max = heights[i].second;
                    
                    if (res.size() > 0 && res.back().first == heights[i].first) {
                        res.back().second = max;
                        if (res.size() > 1 && res[res.size()-2].second == max)
                            res.pop_back();
                    } else res.push_back(make_pair(heights[i].first, max));
                }
                
                ++m[heights[i].second];
                
            } else {
                --m[-heights[i].second];
                if (m[-heights[i].second] == 0)
                    m.erase(-heights[i].second);
                
                if (max == -heights[i].second) {
                    int new_max = 0;
                    if (m.size() > 0)
                        new_max = (--m.end())->first;
                        
                    if (new_max < max) {
                        max = new_max;
                        if (res.size() > 0 && res.back().first == heights[i].first) {
                            if (res.size() > 1 && res[res.size()-2].second > res[res.size()-1].second)
                                res[res.size()-1].second = new_max;
                        } else
                            res.push_back(make_pair(heights[i].first, max));
                    }
                }    
            }
        }
        
        return res;
    }
};

File: java/218TheSkylineProblem.java
class Solution {
    public List<int[]> getSkyline(int[][] buildings) {
        List<int[]> res = new ArrayList<int[]>();
        if (buildings == null)
            return res;
        
        List<int[]> starts = new ArrayList<int[]>();
        List<int[]> ends = new ArrayList<int[]>();
        for (int i = 0; i < buildings.length; ++i) {
            int[] start = new int[2];
            int[] end = new int[2];
            start[0] = buildings[i][0];
            start[1] = buildings[i][2];
            end[0] = buildings[i][1];
            end[1] = buildings[i][2];
            
            starts.add(start);
            ends.add(end);
        }
        
        Comparator comp = new Comparator<int[]>() {
            @Override
            public int compare(int[] first, int[] second) {
                return first[0] - second[0];
            }
        };
        
        Collections.sort(starts, comp);
        Collections.sort(ends, comp);
        
        PriorityQueue<Integer> pq = new PriorityQueue(10, Collections.reverseOrder());
        
        int i = 0, j = 0;
        int prevH = 0;
        while (i < starts.size() && j < ends.size()) {
            int pos = starts.get(i)[0];
            if (pos <= ends.get(j)[0]) {
                while (i < starts.size() && pos == starts.get(i)[0]) {
                    pq.add(starts.get(i)[1]);
                    ++i;
                }
            }
            
            if (pos >= ends.get(j)[0]) {
                pos = ends.get(j)[0];
                while (j < starts.size() && pos == ends.get(j)[0]) {
                    pq.remove(ends.get(j)[1]);
                    ++j;
                }
            }
            
            int height = 0;
            if (!pq.isEmpty()) {
                height = pq.peek();
            }
            if (height != prevH) {
                prevH = height;
                res.add(new int[] {pos, height});
            }
        }
        
        while (j < ends.size()) {
            int pos = ends.get(j)[0];
            while (j < starts.size() && pos == ends.get(j)[0]) {
                pq.remove(ends.get(j)[1]);
                ++j;
            }
            
            int height = 0;
            if (!pq.isEmpty()) {
                height = pq.peek();
            }
            if (height != prevH) {
                prevH = height;
                res.add(new int[] {pos, height});
            }
        }
        
        return res;
    }
}
